jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +183,254 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
+
   # real-valued positive-definite linear operators are symmetric
-  real_valued = lambda x: not issubclass(x.dtype.type, np.complexfloating)
+  def real_valued(x):
+    return not issubclass(x.dtype.dtype, np.complexfloating)
   symmetric = all(map(real_valued, tree_leaves(b)))
   x = lax.custom_linear_solve(
       A, b, solve=cg_solve, transpose_solve=cg_solve, symmetric=symmetric)
   info = None  # TODO(shoyer): return the real iteration count here
   return x, info
+
+
+def _safe_normalize(x, return_norm=False, thresh=None):
+  norm = jnp.sqrt(_vdot_tree(x, x, assume_real=False))
+  if thresh is None:
+    thresh = jnp.finfo(x.dtype).eps
+  thresh = thresh.astype(norm.dtype)
+
+  normalized_x, norm = lax.cond(
+    norm > thresh,
+    lambda y: (_div(y, norm), norm),
+    lambda y: (_mul(0.*y, thresh), thresh),  # To get the dtype right
+    x,
+  )
+  if return_norm:
+    return normalized_x, norm
+  else:
+    return normalized_x
+
+
+def kth_arnoldi_iteration(k, A, M, V, H, tol):
+  """
+  Performs a single (the k'th) step of the Arnoldi process. Thus,
+  adds a new orthonormalized Krylov vector A(M(V[:, k])) to V[:, k+1],
+  and that vectors overlaps with the existing Krylov vectors to
+  H[k, :]. The tolerance 'tol' sets the threshold at which an invariant
+  subspace is declared to have been found, in which case the new
+  vector is taken to be the zero vector.
+  """
+
+  v = tree_map(lambda x: x[..., k], V)  # Gets V[:, k]
+  v = A(M(v))
+
+  def gram_schmidt_step(r, v_i):
+    h_i = _vdot_tree(r, v_i)
+    r_i = _sub(r, _mul(h_i, v_i))
+    return r_i, h_i
+  v, h = lax.scan(gram_schmidt_step, v, xs=V.T)
+  unit_v, v_norm = _safe_normalize(v, return_norm=True, thresh=tol)
+  V = tree_multimap(lambda X, y: X.at[..., k + 1].set(y), V, unit_v)
+  h = h.at[k + 1].set(v_norm)
+  H = H.at[k, :].set(h)
+  return V, H
+
+
+def apply_givens_rotations(H_row, givens, k):
+  """
+  Applies the Givens rotations stored in the vectors cs and sn to the vector
+  H_row. Then constructs and applies a new Givens rotation that eliminates
+  H_row's k'th element.
+  """
+  # This call successively applies each of the
+  # Givens rotations stored in givens[:, :k] to H_col.
+
+  def apply_ith_rotation(i, H_row):
+    cs, sn = givens[i, :]
+    H_i = cs * H_row[i] - sn * H_row[i + 1]
+    H_ip1 = sn * H_row[i] + cs * H_row[i + 1]
+    H_row = H_row.at[i].set(H_i)
+    H_row = H_row.at[i + 1].set(H_ip1)
+    return H_row
+
+  R_row = lax.fori_loop(0, k, apply_ith_rotation, H_row)
+
+  def givens_rotation(v1, v2):
+    t = jnp.sqrt(v1**2 + v2**2)
+    cs = v1 / t
+    sn = -v2 / t
+    return cs, sn
+  givens_factors = givens_rotation(R_row[k], R_row[k + 1])
+  givens = givens.at[k, :].set(givens_factors)
+  cs_k, sn_k = givens_factors
+
+  R_row = R_row.at[k].set(cs_k * R_row[k] - sn_k * R_row[k + 1])
+  R_row = R_row.at[k + 1].set(0.)
+  return R_row, givens
+
+
+def _gmres(A, b, x0, tol, n_kry, M):
```

Comment:
I don't love the `n_kry` name, which took me a while to figure out. Could you stick with the name `maxiter`, or maybe something more verbose like `n_krylov` or `krylov_dim`?

Reply:
I think the relevant parameter is actually very confusingly named `restart`, which is why I favour mentioning the Krylov dimension explicitly. However, I do support either of your two options - perhaps `krylov_dim` slightly more.